### ISSUE TITLE
ld-find-code-refs fails closed on API failures

### ISSUE THREAD

**Issue by garthk** [launchdarkly/ld-find-code-refs#103]
Build broken by `ld-find-code-refs` 1.2.0–1.3.1:
```
$ .bin/ld-find-code-refs -accessToken=$LAUNCH_DARKLY_ACCESS_TOKEN -branch $GIT_BRANCH -contextLines=2 -dir=. -projKey=default -repoName=$REPO -repoType=github -repoUrl=https://github.com/$ORG/$REPO
ld-find-code-refs version 1.3.1
INFO: 2019/10/29 06:48:47 coderefs.go:63: absolute directory path: /home/ci/$REPO
INFO: 2019/10/29 06:48:47 git.go:39: git branch: fix/xxx-yyy-zzz-#999999
INFO: 2019/10/29 06:48:48 search.go:87: finding code references with delimiters: ["'`]
INFO: 2019/10/29 06:48:48 coderefs.go:170: sending 7 code references across 2 flags and 3 files to LaunchDarkly for project: default
FATAL Please file an issue at https://github.com/launchdarkly/ld-find-code-refs: 2019/10/29 06:48:48 coderefs.go:183: error sending code references to LaunchDarkly: invalid_request, field 'name' must be empty or equal to path parameter 'branch'
```

**Comment by atrakh** [comment]
Hi @garthk. Thank you for bringing this to our attention. We've identified a regression in our application causing the branch-updating api route to incorrectly return an `invalid_request` status.

We're working on a fix for this now-- I'll follow-up on this issue once the fix has been deployed. Since this is an issue in our application, existing installations of `ld-find-code-refs` will start working again once the fix has been released, without requiring an upgrade.

**Comment by garthk** [comment]
@atrakh, assuming there'll always be the possibility of back end failure, could you arrange for ld-find-code-refs to \[be able to] exit with status 0 on failures behind the API server? If the API server is in good enough shape to reply with a well formed error other than an authentication or access control failure, perhaps we don't need our build broken. Not having this commit's lines updated is a mild inconvenience. Breaking our build wasn't mild.

**Comment by atrakh** [comment]
Following up on this- a fix has been released and deployed. You should no longer see this error.

Thanks for the feedback, having a build break due to an API failure on our end is certainly not pleasant. Exiting with status 0 could be one way of making sure your workflows don't break in these cases. I'm hesitant to do that by default, as it could result in silent failures in this cass. We'll think through this for our next release `ld-find-code-refs` release.

Can you tell me more about your build system that's running `ld-find-code-refs`? We'd typically consider code references to be a non-critical build step, and recommend including it in a separate non-blocking workflow when possible.

**Comment by garthk** [comment]
Assuming no behaviour change:

* From outside `ld-find-code-refs`, we can mask failure with an `|| true` in our script… but then if we broke our configuration, or the command line syntax, then `ld-find-code-refs` wouldn't be working for a long while and we just wouldn't know.

* If we don't mask failure, any transient LD API problem will break our build.

* If we put it in another part of the build pipeline as you suggest, we'll still have the choice between either risking our build breaking or risking not knowing that the integration was broken. 

That's a brittle outcome, which suggests to me that one of our assumptions might be wrong. In this case, I think it's the assumption that a command line tool must always exit >0 whenever anything goes wrong, ever. That's generally true… until you start dealing with distributed systems, which we've learned take special care e.g. the [circuit breaker pattern](https://martinfowler.com/bliki/CircuitBreaker.html).

I reckon `ld-find-code-refs` should have an option or default to exit 0 _for the case of API failures only_. Command line malformed? Exit with an error. Bad token? Exit with an error. DNS problem? Exit with an error. But, successfully authenticated and got a well formed response telling us that LD wasn't available for a chat right now? Exit 0.

**Comment by atrakh** [comment]
Hi @garthk, we've released [1.4.0](https://github.com/launchdarkly/ld-find-code-refs/releases/tag/1.4.0), which adds an `ignoreServiceErrors` option. When enabled, transient issues in the LaunchDarkly API will still cause ld-find-code-refs to terminate, just with a 0 exit code.
